### PR TITLE
Split up NotebookProvider into separate providers for handling file serialization and cell execution.

### PR DESCRIPTION
**Pull request by corivera**
As part of the work to enable VSCode's notebook APIs in ADS, I needed to split up our NotebookProvider interface into separate interfaces for handling file serialization and cell execution. This is because VSCode's notebooks allow users to register those types of behaviors seprately, so we can't rely on extensions registering everything all at once if we want to support the VSCode behavior. The general changes in interfaces looks like the following:

Previous interfaces:
A NotebookProvider which contains a NotebookManager, which in turns contains a ContentManager, a SessionManager, and a ServerManager.

New interfaces:
1. For cell execution, we now have a NotebookExecuteProvider, which returns an ExecuteManager, which contains a SessionManager and ServerManager
2. For file serialization and deserialization, we now have a NotebookSerializationProvider, which returns a SerializationManager, which contains a ContentManager.
Note: The reason we have a wrapper interface in the form of SerializationManager has to do with how the notebook code passes around these registration objects in different ways internally. Just having the Provider interfaces provide their lowest level managers caused some headaches when trying to refactor the code, so I kept the NotebookManager "middleman" concept for these new APIs to reduce the amount of already very large code churn.

These changes only affected the registration APIs in azdata.d.ts, while the actual provider registration in the package.json was unaffected (i.e. how we set things like providerId, standardKernels, and supported file extensions). This is now represented by a "ProviderDescriptionRegistration" object in the NotebookRegistry. The rest of the notebook code queries the NotebookRegistry for this top level metadata, and then correlates it later with the contributed ExecuteProviders and SerializationProviders using their provider IDs.

Additionally, the ContentManager interface was refactored to only handle converting back and forth from string file contents and our JSON notebook representation (INotebookContents). Before there was an assumption about having to do file I/O as part of this operation, but file I/O is instead handled by the underlying editor model, rather than our manager objects.

ContentManager contributions (via NotebookSerializationProviders) were also actually enabled as part of this work, since before we didn't actually honor any extension registrations for ContentManagers. We just ignored it and used our own internal LocalContentManager class instead, and didn't even use the actual registration APIs to do so. It was just initialized internally when no ContentManager was set, which would never happen since we didn't actually do anything with the ContentManagers that were passed in. I verified that this worked by creating a dummy NotebookSerializationProvider for handling CSV "notebooks" rather than JSON ones, and our notebooks were able to correctly load and save with an underlying comma separated format.

Apart from those details, I took some liberties in simplifying or renaming certain things for clarity. These broadly fall into the following categories:
1. Removed deprecated standardKernels field from defunct NotebookProvider
2. Renamed IContentManager interface in modelInterfaces.ts to IContentLoader, since it's unrelated to azdata.nb.ContentManager
3. Added new SqlSerializationProvider class to register our default LocalContentManager, just like with we do with our NotebookProviders (now NotebookExecuteProviders)
4. Updated various parts of the code to actually pull info from the current notebook provider, rather than using hardcoded provider IDs (like jupyter or sql) or hardcoded file extensions (like .ipynb)
5. Removed unused remoteContentManager.ts file and its associated contentManagers.test.ts. Neither of these was actually referenced in our code, and their last modifications happened 2 years ago.
6. Reduced time spent waiting for Notebook provider registration in the NotebookService from 30 seconds to 10 seconds. This is because we now wait on ExecuteProvider and SerializationProvider registrations separately, but either of those might not be contributed by extensions, which led to long wait times in notebook start up in testing (i.e. waiting for kernels to finish loading). I also changed these registration waits to remove the underlying awaited "ProviderDescriptor" object on timeout, since we shouldn't keep waiting for these non-existent providers to load each time we open a notebook.

Apart from these details, much of the changes in this PR are effectively renames. Most of the time anything that was a NotebookProvider or NotebookManager before is now a NotebookExecuteProvider or ExecuteManager instead. We only really care about NotebookSerializationProviders and SerializationManagers in a small number of places. The files that had meaningful changes besides renames or test changes broadly fall into this list:
1. azdata.d.ts
2. mainThreadNotebook.ts
3. extHostNotebook.ts
4. notebookServiceImpl.ts
5. notebookInput.ts
6. notebookTextFileModel.ts
7. notebookEditor.component.ts
8. notebookRegistry.ts
9. mainThreadEditors.ts

### CODE REVIEW

File: src/vs/workbench/test/browser/api/extHostNotebookKernel2.test.ts [hunk 1]
```diff
@@ -65,8 +65,8 @@ suite('NotebookKernel', function () {
 			override async $applyEdits() { }
 		});
 		rpcProtocol.set(MainContext.MainThreadNotebook, new class extends mock<MainThreadNotebookShape>() {
-			override async $registerNotebookProvider() { }
-			override async $unregisterNotebookProvider() { }
+			override async $registerContentProvider() { }
```
Comment (chlafreniere) [inline src/vs/workbench/test/browser/api/extHostNotebookKernel2.test.ts:68]: Do we have to change these files in src/vs?

File: src/sql/workbench/test/electron-browser/api/mainThreadNotebook.test.ts [hunk 2]
```diff
@@ -57,88 +57,134 @@ suite('MainThreadNotebook Tests', () => {
...
-			let manager = await provider.getNotebookManager(notebookUri);
+			let manager = await serializationProvider.getSerializationManager(notebookUri);
 
-			// Then it shouldn't have wrappers for the content or server manager
+			// Then it shouldn't have wrappers for the content manager
 			assert.ok(!(manager.contentManager instanceof LocalContentManager));
+		});
+
+		test('should return execute manager with a server manager if extension host has these', async () => {
+			// Given the extension provider doesn't have acontent or server manager
```
Comment (lucyzhang929) [inline src/sql/workbench/test/electron-browser/api/mainThreadNotebook.test.ts:169]: typo: "a content"

File: src/vs/workbench/api/browser/mainThreadEditors.ts [hunk 3]
```diff
@@ -136,14 +138,16 @@ export class MainThreadTextEditors implements MainThreadTextEditorsShape {
 	async $tryShowTextDocument(resource: UriComponents, options: ITextDocumentShowOptions): Promise<string | undefined> {
 		const uri = URI.revive(resource);
 
+		let notebookFileTypes = this._notebookService.getSupportedFileExtensions().map(s => s.toLowerCase());
```
Comment (lucyzhang929) [inline src/vs/workbench/api/browser/mainThreadEditors.ts:141]: What is the general rule around editing vs files? Does each line need a {{SQL CARBON EDIT}}?

File: src/sql/workbench/test/electron-browser/api/exthostNotebook.test.ts [hunk 4]
```diff
@@ -13,132 +13,213 @@ import { IMainContext } from 'vs/workbench/api/common/extHost.protocol';
...
-				let managerDetails: INotebookManagerDetails = await extHostNotebook.$getNotebookManager(providerHandle, notebookUri);
+				let managerDetails: ISerializationManagerDetails = await extHostNotebook.$getSerializationManagerDetails(serializationProviderHandle, notebookUri);
 
 				// Then I expect the same manager to be returned
 				assert.ok(managerDetails.hasContentManager === false, 'Expect no content manager defined');
+				assert.ok(managerDetails.handle > 0, 'Expect a valid handle defined');
+			});
+
+			test('Should return an execute manager with correct info on server manager existence', async () => {
+				// Given the provider returns a manager with no
```
Comment (lucyzhang929) [inline src/sql/workbench/test/electron-browser/api/exthostNotebook.test.ts:94]: Was this comment cut off?

File: src/sql/workbench/contrib/notebook/test/browser/notebookInput.test.ts [hunk 5]
```diff
@@ -87,11 +93,11 @@ suite('Notebook Input', function (): void {
 		// Notebook URI
 		assert.deepStrictEqual(untitledNotebookInput.notebookUri, untitledUri);
 
-		// Content Manager
+		// Notebook editor timestamp
 		assert.notStrictEqual(untitledNotebookInput.editorOpenedTimestamp, undefined);
 
-		// Notebook editor timestamp
-		assert.notStrictEqual(untitledNotebookInput.contentManager, undefined);
+		// Content Manager
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/browser/notebookInput.test.ts:99]: "Content Loader"?

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookEditorModel.test.ts [hunk 6]
```diff
@@ -50,7 +50,12 @@ class ServiceAccessor {
...
 
-class NotebookManagerStub implements INotebookManager {
+class SerializationManagerStub implements ISerializationManager {
+	providerId: string;
+	contentManager: nb.ContentManager;
+}
+
+class ExecuteManagerStub implements IExecuteManager {
 	providerId: string;
 	contentManager: nb.ContentManager;
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/electron-browser/notebookEditorModel.test.ts:60]: Should this be serverManager?

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 7]
```diff
@@ -135,27 +135,43 @@ export class NotebookModel extends Disposable implements INotebookModel {
...
-		let manager = this.notebookManagers.find(manager => manager.providerId === this._providerId);
+	public get serializationManager(): ISerializationManager | undefined {
+		let manager = this.serializationManagers.find(manager => manager.providerId === this._providerId);
+		if (!manager) {
+			manager = this.serializationManagers.find(manager => manager.providerId === DEFAULT_NOTEBOOK_PROVIDER);
+		}
+		return manager;
+	}
+
+	public get executeManagers(): IExecuteManager[] {
```
Comment (lucyzhang929) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:154]: Should this be private like `get serializationManagers()`?

File: src/sql/workbench/api/common/extHostNotebook.ts [hunk 8]
```diff
@@ -272,23 +312,39 @@ export class ExtHostNotebook implements ExtHostNotebookShape {
...
 		}
-		return this.callbackWithErrorWrap<R>(callback, manager);
+		return this.callbackWithErrorWrap<SerializationManagerAdapter, R>(callback, manager);
 	}
 
-	private async callbackWithErrorWrap<R>(callback: (manager: NotebookManagerAdapter) => R | PromiseLike<R>, manager: NotebookManagerAdapter): Promise<R> {
+	private _withExecuteManager<R>(handle: number, callback: (manager: ExecuteManagerAdapter) => R | PromiseLike<R>): Promise<R> {
+		let manager = this._adapters.get(handle) as ExecuteManagerAdapter;
+		if (manager === undefined) {
+			return Promise.reject(new Error(localize('errNoManager', "No Manager found")));
```
Comment (lucyzhang929) [inline src/sql/workbench/api/common/extHostNotebook.ts:342]: Do we want to specify "No Execute Manager found" here?

**Review by lucyzhang929 — APPROVED** [review]

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts (hunk 7 above)
Comment (corivera) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:154]: No, this gets used in a couple of other places.

**Pull request merged**